File: implementations/julia/src/rng.jl
const PCG32_MULTIPLIER = UInt64(6364136223846793005)

mutable struct PCG32
    state::UInt64
    increment::UInt64

    function PCG32(seed::Integer, stream::Integer = 54)
        seed >= 0 || throw(ArgumentError("seed must be non-negative"))
        stream >= 0 || throw(ArgumentError("stream must be non-negative"))
        rng = new(UInt64(0), (UInt64(stream) << 1) | UInt64(1))
        next_uint32!(rng)
        rng.state += UInt64(seed)
        next_uint32!(rng)
        return rng
    end
end

function next_uint32!(rng::PCG32)::UInt32
    old = rng.state
    rng.state = old * PCG32_MULTIPLIER + rng.increment
    xorshifted = UInt32((((old >> 18) ⊻ old) >> 27) & UInt64(0xffffffff))
    rotation = Int(old >> 59)
    return (xorshifted >> rotation) | (xorshifted << ((-rotation) & 31))
end

function next_float32!(rng::PCG32)::Float32
    return Float32(next_uint32!(rng)) * Float32(1.0 / 4294967296.0)
end

File: implementations/typescript/src/core/rng.ts
const MASK64 = (1n << 64n) - 1n;
const MULTIPLIER = 6364136223846793005n;

export class Pcg32 {
  private state = 0n;
  private readonly increment: bigint;

  public constructor(seed: number, stream = 54) {
    if (!Number.isSafeInteger(seed) || seed < 0 || seed > 0xffff_ffff) throw new RangeError("seed must be uint32");
    if (!Number.isSafeInteger(stream) || stream < 0) throw new RangeError("stream must be nonnegative");
    this.increment = ((BigInt(stream) << 1n) | 1n) & MASK64;
    this.nextUint32();
    this.state = (this.state + BigInt(seed)) & MASK64;
    this.nextUint32();
  }

  public nextUint32(): number {
    const old = this.state;
    this.state = (old * MULTIPLIER + this.increment) & MASK64;
    const xorshifted = Number((((old >> 18n) ^ old) >> 27n) & 0xffff_ffffn) >>> 0;
    const rotation = Number((old >> 59n) & 31n);
    return ((xorshifted >>> rotation) | (xorshifted << ((-rotation) & 31))) >>> 0;
  }

  public nextFloat32(): number {
    return Math.fround(Math.fround(this.nextUint32()) * Math.fround(2 ** -32));
  }

  public checkpoint(): bigint { return this.state; }

  public restore(checkpoint: bigint): void {
    if (checkpoint < 0n || checkpoint > MASK64) throw new RangeError("PCG32 checkpoint must be uint64");
    this.state = checkpoint;
  }

  public fill(target: Float32Array): void {
    for (let index = 0; index < target.length; index += 1) target[index] = this.nextFloat32();
  }
}
